File: aws-s3/src/service/ImageService.ts
import { v4 as uuidv4 } from 'uuid';
import { UserRepository } from '../repository/UserRepository';
import { UserPictureRepository } from '../repository/UserPictureRepository';
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import * as path from 'path';
import * as fs from 'fs';
import { UserPicture } from '../entity/UserPicture';
import { ReadStream } from 'fs';

@Injectable()
export class ImageService {
  constructor(private userRepository: UserRepository, private userPictureRepository: UserPictureRepository) {}
  async saveImage(userId: string, file: Express.Multer.File): Promise<void> {
    const user = await this.userRepository.findById(userId);
    if (!user) {
      throw new NotFoundException('User not found');
    }

    const picture = await this.userPictureRepository.findByUserIdAndName(userId, file.originalname);
    if (picture) {
      throw new BadRequestException('the file already exits for that user');
    }

    const filename = `${Date.now()}-${file.originalname}`;
    const filePath = path.join(__dirname, '..', 'pictures', filename);
    const userPicture = new UserPicture(uuidv4(), user.id, file.originalname, filePath);
    const folderPath = path.dirname(filePath);

    // Create the folder if it doesn't exist
    if (!fs.existsSync(folderPath)) {
      fs.mkdirSync(folderPath, { recursive: true });
    }
    //write to disk
    await fs.promises.writeFile(filePath, file.buffer);

    //save in db
    await this.userPictureRepository.save(userPicture);
  }

  async getImageStream(userId: string, filename: string): Promise<ReadStream> {
    const picture = await this.userPictureRepository.findByUserIdAndName(userId, filename);
    if (!picture) {
      throw new NotFoundException('file not found');
    }

    const filePath = picture.url;

    return fs.createReadStream(filePath);
  }
}

File: aws-s3/spring-boot/src/main/java/com/example/service/ImageService.java
package com.example.service;

import com.example.entity.UserPicture;
import com.example.repository.UserPictureRepository;
import com.example.repository.UserRepository;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

@Service
@RequiredArgsConstructor
public class ImageService {

    private final UserRepository userRepository;
    private final UserPictureRepository userPictureRepository;

    @SneakyThrows
    public void saveImage(String userId, MultipartFile file) {
        var user =
                userRepository
                        .findById(userId)
                        .orElseThrow(() -> new RuntimeException("User not found"));
        userPictureRepository
                .findByUserIdAndName(userId, file.getOriginalFilename())
                .ifPresent(
                        (picture) -> new RuntimeException("The file already exists for that user"));

        String filename = UUID.randomUUID() + "-" + file.getOriginalFilename();
        Path filePath = Paths.get("pictures", filename);
        UserPicture userPicture =
                new UserPicture(
                        UUID.randomUUID().toString(),
                        user.getId(),
                        file.getOriginalFilename(),
                        filePath.toString());

        Files.createDirectories(filePath.getParent());
        Files.write(filePath, file.getBytes());
        userPictureRepository.save(userPicture);
    }

    @SneakyThrows
    public Resource getImage(String userId, String filename) {
        var picture =
                userPictureRepository
                        .findByUserIdAndName(userId, filename)
                        .orElseThrow(() -> new RuntimeException("File not found"));

        Path filePath = Paths.get(picture.getUrl());
        Resource resource = new UrlResource(filePath.toUri());

        if (resource.exists() || resource.isReadable()) {
            return resource;
        } else {
            throw new RuntimeException("Could not read the file!");
        }
    }
}
